File: src/ios/Classes/Pipeline/PipelineStage.swift
import Foundation

/// In-place processing stage. No allocation, no blocking, no Flutter calls.
/// Stages should not throw under normal operation — PipelineRunner catches
/// any exception and disables the stage after repeated failures.
protocol PipelineStage: AnyObject {
    func initialize(config: PipelineConfig)
    func process(frame: MutableAudioFrame) throws
    func start()
    func stop()
    func dispose()
}

File: src/android/src/main/kotlin/com/circuids/boat/pipeline/PipelineStage.kt
package com.circuids.boat.pipeline

/**
 * In-place processing stage. No allocation, no return value, never throws.
 */
interface PipelineStage {
    fun initialize(config: PipelineConfig)
    fun process(frame: MutableAudioFrame)
    fun start()
    fun stop()
    fun dispose()
}
